### DISCUSSION TITLE
Loaded TTF does not draw properly on Ubuntu MATE for Raspberry Pi aarch64

### DISCUSSION THREAD

**Discussion by ghost** [ocornut/imgui#4987] [New Users Build/Link/Run issues ONLY!]
likely happens on armhf and will test on there if anyone needs me to confirm. The same exact code displays fonts properly on OpenBSD x86-64 and macOS x86-64. My code uses OpenGL2 because OpenGL3 is not compatible. I'm using the raspberry pi 4 model b. I tried the OpenGL3 equivalent on platforms where it is compatible, including macOS x86-64 (again), Windows x86-64, Windows x86, FreeBSD x86-64, FreeBSD aarch64 (Pinebook Pro img from sleepwalker on the freebsd forum), and also ubuntu 20.04 LTS (the x86-64 GNOME edition) and Manjaro. I am using 20.04 LTS ubuntu MATE for the pi more specifically for the screenshot produced below, literally the only platform so far that doesn't render correctly with the same exact code as my other tested platforms: 

![Screenshot at 2022-02-06 00-19-27](https://user-images.githubusercontent.com/4379204/152668737-d943ff02-0709-4317-8d5b-bcb7aa86059e.png)

Relevant OpenGL3 code here: https://github.com/time-killer-games/ImGuiFileDialog-SDL2-OpenGL3/blob/048691d423e2c520d08685e2c7228aa63f152723/filedialogs.cpp#L356-L484

And here's the OpenGL2 code i need since OpenGL3 isn't compatible but this one renders the fonts wrong on the platform mentioned:
https://github.com/time-killer-games/ImGuiFileDialog-SDL2-OpenGL2/blob/42257ae653b436d1daaa246f28b3f5889fb2032a/filedialogs.cpp#L356-L465

These are taken very directly from the example code for OpenGL 2, 3, and SDL2 in the imgui repository. 

Edit: also happens on Raspberry Pi OS armhf. not sure what's making the difference.

**Comment by thedmd** [comment]
`GL_RGBA` internal format may not ba available. Try `GL_BGRA`. If you're using OpenGL ES then pass it as both texture and internal texture format. If that works you may have to swap r/b channels manually to get correct colors, if your font atlas does not contain colored emoji. That may be not be an issue, as grayscale image will look same in both cases.

**Reply by ghost** [reply]
Thank you for the insight! I'll play around with the things you mentioned a bit and see what happens. I'll mark this as solved or ask more questions depending on the outcome.

**Reply by ghost** [reply]
I tried `GL_BGRA` for the internal format argument:
`glTexImage2D(GL_TEXTURE_2D, 0, GL_BGRA, width, height, 0, GL_RGBA, GL_UNSIGNED_BYTE, pixels);`

I tried `GL_BGRA` for the internal format and "regular" format arguments, compiled with `-DIMGUI_IMPL_OPENGL_ES2`:
`glTexImage2D(GL_TEXTURE_2D, 0, GL_BGRA, width, height, 0, GL_BGRA, GL_UNSIGNED_BYTE, pixels);`

...and I had the same result with the fonts lacking transparency for both, thus it didn't fix the issue.

However it wouldn't surprise me if that's because I did something wrong. It's hard to say.

**Reply by ghost** [reply]
Also it's worth noting I didn't add that line, I modified the existing call to `glTexImage2D()` under `ImGui_ImplOpenGL2_CreateFontsTexture()`, which if I read correctly, I think that's what you wanted me to do.

**Comment by ghost** [comment]
@thedmd i found out the issue was i was loading way more fonts than the raspberry pi and similar ARM devices can handle all at once apparently. on x86 PC's i never have these issues compiled using the same code. However if i reduce the amount of font files significantly i attempt to load/combine for multilingual support, the issue goes away. For example, trying to load one font file instead of hundreds of them.

I have an example program added to the readme if to make testing easier, or just run the build script as-is. Either way...
https://github.com/time-killer-games/libfiledialogs/blob/main/README.md

**Reply by andreamancuso** [reply]
Thank you, I was stuck with this, this is now working for me on my Raspberry Pi 5!
